src/libs/Middleware/SaveResponseInOnyx.js

```diff
@@ -9,21 +9,29 @@ import _ from 'underscore';
 function SaveResponseInOnyx(response, request) {
     return response
         .then((responseData) => {
-            // We'll only save the onyxData, successData and failureData for the refactored commands
-            if (_.has(responseData, 'onyxData')) {
-                const data = [];
-                if (responseData.jsonCode === 200) {
-                    if (request.successData) {
-                        data.push(...request.successData);
-                    }
-                } else if (request.failureData) {
-                    data.push(...request.failureData);
-                }
-                if (responseData.onyxData) {
-                    data.push(...responseData.onyxData);
-                }
-                Onyx.update(data);
+            const onyxUpdates = [];
+
+            // Make sure we have response data (i.e. response isn't a promise being passed down to us by a failed retry request and responseData undefined)
+            if (!responseData) {
+                return;
+            }
+
+            // Handle the request's success/failure data (client-side data)
+            if (responseData.jsonCode === 200 && request.successData) {
+                onyxUpdates.push(...request.successData);
+            } else if (responseData.jsonCode !== 200 && request.failureData) {
+                onyxUpdates.push(...request.failureData);
+            }
```

Comment:
cc @yuwenmemon 

I'm not sure if this case matters, but if a requests hits the back end with a command that is not implemented, the backend returns a 404. I would have expected that the `failureData` would have pushed into Onyx, but it isn't because `responseData` is undefined so I ended up without feedback. Could this happen in a real case? like if we try to send a request about a workspace or a chat that doesn't exist anymore?

Reply:
I see what you mean because the response for that case does not come via `jsonCode`: https://github.com/Expensify/Web-Expensify/blob/30edfabd582f9ac70ec59cef3172a5bf25657efa/api.php#L2761-L2766


Maybe we can add a parameter that indicates if this call to the Middleware is coming from a failed retry request or not? Maybe checking the [retryCount](https://github.com/Expensify/App/blob/d238b2368357f8954f1fb290966975a9c3792572/src/libs/Middleware/Retry.js#L20)? Or check the HTTP response code somehow? The original issue was that if a request was a failed retry (i.e. could not connect to the Server), we'd have no `responseData` and thus get an undefined error. 

**However** I don't really think this case matters too much right? Why would we want to support error messages for non-existent API commands? We shouldn't be calling non-existent API commands in the production App ever, right? Any "feedback in the UI" would solely be serving developers and that's it, no?

Reply:
> However I don't really think this case matters too much right? Why would we want to support error messages for non-existent API commands? We shouldn't be calling non-existent API commands in the production App ever, right? Any "feedback in the UI" would solely be serving developers and that's it, no?

I completely agree that this doesn't matter if it only happens when a command doesn't exist, but I had my doubt if this could happen with other crashes like and exception being thrown. I'll test and let you know.

Reply:
Ok, retested throwing an exception in the web API, we get a `jsonCode = 666`, `responseData` is not undefined, and the `failureData` is applied correctly. I think this is the case that matters and works fine.. so lets do nothing :)